### ISSUE TITLE
Wrap long addr:housenumber labels onto multiple lines

### ISSUE THREAD

**Issue by 1ec5** [gravitystorm/openstreetmap-carto#4155]
### Expected behavior

Multiple values in `addr:housenumber` should ideally be rendered in a multi-line grid, with semicolons or commas replaced with whitespace, similar to how multiple values in `ref` are rendered as a multiline shield. If that isn’t feasible, at least long `addr:housenumber` labels should be line-wrapped.

### Actual behavior

`addr:housenumber` with many values extends far beyond the building, looking unwieldy. This behavior encourages mappers to move the `addr:housenumber` to nodes laid out in an arbitrary grid within the building, even if the actual layout within the building is unknown.

### Links and screenshots illustrating the problem

![willowbrook](https://user-images.githubusercontent.com/1231218/83653451-a9170600-a570-11ea-92c2-331cd50cc504.png)

**Comment by jeisenbe** [comment]
I suspect that this is non-standard tagging. Normally the key `addr:housenumber` has only one number in the value. Are these unit numbers perhaps? Could you provide a link to the location of the screenshots.

To map a large range of housenumbers, for example an apartment building with a number of different flats, it is common to use a dash to show the range, e.g. 25-63. This is most often done with [addr:flats](https://wiki.openstreetmap.org/wiki/Key%3Aaddr%3Aflats) which we just started rendering recently: 

"Use addr:flats=* for marking range of numbers of flats behind a door, for example, addr:flats=1-20. If there is gap in numeration, mark numbers and intervals with semicolon: addr:flats=3-7;10;14;16-18."

This suggests that we could add support for semicolons in addr:flats. I suppose we could also consider this for addr:housenumber (a search for ";" in taginfo finds a number of uses: https://taginfo.openstreetmap.org/keys/?key=addr%3Ahousenumber#values) but it would not necessarily fix the issue with the areas that are shown above.

**Comment by kennykb** [comment]
The first screenshot is one that I tagged and happened to mention on one of the Slack channels. Minh has shared it here.  

https://www.openstreetmap.org/way/491160756 is an example.  While there's a communal entrance to the buiilding, the individual units have addresses '101 Connor Court', '102 Connor Court', etc., exactly as if they were detached houses or terraces.  This contrasts with a similar building a short distance away, https://www.openstreetmap.org/way/491155947, where the addresses are '35 Hillcrest Village East, apartment 1A' and so on.  If you insist that flats behind a door need to be addr:units, then there's nothing to inform the correct syntax for addressing them, and a search for '102 Connor Court' in a navigation system will fail.

The building's door has '101 102 103 104' down one doorpost, and '105 106 107 108' down the other, suggesting that perhaps one turns left for one set of units and right for the other, but I had no business inside to determine how they're laid out. These, for all intents and purposes, are house numbers.

If they had separate entrances, I'd have attached the numbers to them, like https://www.openstreetmap.org/way/554579973. But what I can see from the street on Connor Court  is series of buildings, each of which sports a single front door with eight house numbers posted on it.

**Comment by Thorwynn** [comment]
I recognize that 2nd screenshot from Hull, Massachusetts (I helped Alan Bragg with fixing street name errors during the address import and remembered fixing streets in that area) - it is https://www.openstreetmap.org/way/213653111/ if you look at the Bing imagery & streetside it is crazy complicated and only 2 doors appear to be visible from the outside.

**Comment by 1ec5** [comment]
Some particularly egregious examples at z17 in San José, California (including [this building](https://www.openstreetmap.org/relation/11922056):

![sanjose](https://user-images.githubusercontent.com/1231218/123510138-aa5d1a00-d62e-11eb-8ebf-bcd42e89f253.png)

**Comment by imagico** [comment]
A possible solution to this would be to replace long addr:housenumber values of the form `A;B;C;D;E` with `A...E`.

**Comment by 1ec5** [comment]
An ellipsis makes a lot of sense. Showing the ending value (“E”) after the ellipsis would assume the list of values is already sorted in numerical or alphabetical order, which might not necessarily be the case. I tend to tag such lists in the same order as on the sign. However, I think that’s a very minor issue compared to the unwieldiness of these long labels.

Incidentally, I was wondering if it’s also possible to allow shorter address labels to collide out longer ones. The idea would be to show more buildings’ addresses, rather than allowing a few buildings to dominate a neighborhood as in the San José example.

**Comment by imagico** [comment]
Yes, you could prioritize by labeling string length.  You would have to assemble the labeling string in SQL for that (which is currently done in MSS).  But you might end up showing the addresses of small buildings (which just have the housenumber) while not showing those of larger blocks with long labeling strings due to multiple housenumbers and/or unit/flats numbers.  So abbreviation of long strings plus separating unit/flats labels and showing them with lower priority (see https://github.com/gravitystorm/openstreetmap-carto/issues/4160#issuecomment-868987197) could be worth trying.

**Comment by ZeLonewolf** [comment]
Here is another example I found in the wild in Fall River, Massachusetts, USA, with other long address strings nearby out of frame ([location](https://www.openstreetmap.org/#map=18/41.69952/-71.17204)):

![image](https://github.com/gravitystorm/openstreetmap-carto/assets/3254090/307bf301-ba14-498e-971d-c6c237d911e0)

**Comment by Circeus** [comment]
Even if tehse were reduced to ranges, it it not difficult to find duplexes in montreal where two numbers hyphenated give a total of eight or even 10 digit.
![image](https://github.com/gravitystorm/openstreetmap-carto/assets/6731840/fc4c2850-2a6f-4278-8e2a-ca2939c23e0c)
And these are large buildings. usually these addresses are found on much smaller buildings like those here:
![image](https://github.com/gravitystorm/openstreetmap-carto/assets/6731840/d149e0aa-d6e1-4170-8b44-4aaa21d079ac)

**Comment by dch0ph** [comment]
> An ellipsis makes a lot of sense. Showing the ending value (“E”) after the ellipsis would assume the list of values is already sorted in numerical or alphabetical order, which might not necessarily be the case. I tend to tag such lists in the same order as on the sign. However, I think that’s a very minor issue compared to the unwieldiness of these long labels.

Shortening `A;B;C;D` to `A...D` seems a good solution. I don't see that it implies that values are sorted. It just indicates that "there are too many values to show without cluttering the map". 

The `;` does make it fair game for re-writing - it is an internal marker of multiple values. You could even re-write any use of ; in `housenumber` or `flats`:
`A;B` -> `A,B`
`A;B;C` (and above) -> `A...C`
